File: Array/Medium/python/2sum.py
#brute force
def twoSum(arr, k):
    for i in range(len(arr)-1):
        for j in range (i+1,len(arr)):
            if arr[i]+arr[j] == k:
                return [i,j]
    return [-1,-1]

arr = [2,6,5,8,11]
k=15
print(twoSum(arr,k))


#using hashing
def twoSum(arr, k):
    hashdata= {}
    for i in range(len(arr)):
        rem = k - arr[i]
        if rem in hashdata:
            return [hashdata[rem], i]
        else:
           hashdata[arr[i]] = i
    return [-1,-1]

arr = [2,6,5,8,11]
k=14
print(twoSum(arr,k))

#two pointer: but for this approach, the array needs to be sorted
def twoSum(arr, k):
    i = 0
    j = len(arr)-1
    while(i<j):
        sum = arr[i] + arr[j]
        if k > sum:
            i=i+1
        elif k < sum:
            j=j-1
        else:
            return [i,j]
    return [-1,-1]

arr = [2,6,5,8,11]
k=17
print(twoSum(arr,k))

File: Array/Medium/js/2sum.js
/** Two Sum : Check if a pair with given sum exists in Array */
arr = [2,6,5,8,11]
target = 14
/** 
 * Brute Force
 * TC: O(N)^2
 * SC: O(1)
 *  */
function calculate2Sum(arr, target){
    let N = arr.length
    for (let i=0;i<N;i++){
        for( let j=i;j<N;j++){
            if((arr[i] + arr[j]) == target){
                return `YES : ${arr[i]},${arr[j]}`
            }
        }
    }
    return 'NO'
}
console.log('Brute Froce:: ',calculate2Sum(arr,target))


/** 
 * Using hashing & reverse mathematics
 * Calculating the remaining = target - currentElement, if that exist in hashMap means we got 2 sum
 * else store the current element in hashmap {element: key} 
 * SC: O(N)
 * TC: O(N)
 *  */
function calculate2SumOp(arr, target){
    let hashMap = {}
    for(let i=0;i<arr.length;i++){
        let rem = target - arr[i]
        if(hashMap[rem]){
            return `YES : ${rem},${arr[i]}`
        }
        hashMap[arr[i]] = i
    }
    console.log(hashMap)
    return 'NO'
}
console.log('Using Hashing :: ',calculate2SumOp(arr,target))


/** 
 * Using 2 pointers
 * Required the array to be sorted. Here we are sorting the array first
 * TC: O(N)
 * SC: O(NlogN)
 *  */

function calculate2Sum2P(arr,target){
    let N = arr.length
    let left = 0;
    let right = N-1;
    arr = arr.sort((a,b) => a-b)

    while(left < right ){
        let sum = arr[left] + arr[right]
        if (sum == target){
            return `YES : ${arr[left], arr[right]}`
        } else if (sum > target){
            right--
        } else {
            left++
        }
    }
    return 'NO';
}
console.log('Using 2 pointers :: ',calculate2SumOp(arr,target))